File: test/core.test.ts
import { describe, expect, it } from "vitest";
import { createStepFn, createWorkflow, ok, type StepFn } from "../src/core.js";
import { EXIT, ERR, STEPS_TRACE } from "../src/keys.js";

const aStepFn: StepFn = (f, step, opts) => {
  const started = { ...opts, [STEPS_TRACE]: [...(opts[STEPS_TRACE] ?? []), [step, "start-a"]] };
  const next = f(step, started);
  return { ...next, [STEPS_TRACE]: [...(next[STEPS_TRACE] ?? []), [step, "end-a"]] };
};

const bStepFn: StepFn = (f, step, opts) => {
  const started = { ...opts, [STEPS_TRACE]: [...(opts[STEPS_TRACE] ?? []), [step, "start-b"]] };
  const next = f(step, started);
  return { ...next, [STEPS_TRACE]: [...(next[STEPS_TRACE] ?? []), [step, "end-b"]] };
};

describe("core", () => {
  it("validates step-fn options", () => {
    expect(() => createStepFn({})).toThrow(/At least one/);
    expect(() => createStepFn({ afterF: "same" })).toThrow(/before-f/);
  });

  it("threads opts and wraps step-fns in order", () => {
    const wf = createWorkflow({
      firstStep: "test/start",
      wireFn: (step) => {
        switch (step) {
          case "test/start": return [(opts) => ({ ...ok(opts), "test/bar": "baz" }), "test/end"];
          case "test/end": return [(opts) => opts];
          default: return [(opts) => opts];
        }
      }
    });
    const res = wf([aStepFn, bStepFn], {});
    expect(res[EXIT]).toBe(0);
    expect(res[ERR]).toBeNull();
    expect(res["test/bar"]).toBe("baz");
    expect(res[STEPS_TRACE]).toEqual([
      ["test/start", "start-a"],
      ["test/start", "start-b"],
      ["test/start", "end-b"],
      ["test/start", "end-a"],
      ["test/end", "start-a"],
      ["test/end", "start-b"],
      ["test/end", "end-b"],
      ["test/end", "end-a"]
    ]);
  });

  it("captures exceptions into opts", () => {
    const wf = createWorkflow({
      firstStep: "test/start",
      wireFn: (step) => step === "test/start"
        ? [() => { throw Object.assign(new Error("boom"), { data: { extra: 1 } }); }, "test/end"]
        : [(opts) => opts]
    });
    const res = wf([], {});
    expect(res[EXIT]).toBe(1);
    expect(res[ERR]).toBe("boom");
    expect(res.extra).toBe(1);
  });
});

File: test/big_config/core_test.clj
(ns big-config.core-test
  (:require [big-config.core :as core]
            [big-config.keys :as k]
            [clojure.test :refer [deftest is testing]]))

(defn a-step-fn [f step opts]
  (let [started (assoc opts k/steps-trace (conj (vec (get opts k/steps-trace [])) [step :start-a]))
        next (f step started)]
    (assoc next k/steps-trace (conj (vec (get next k/steps-trace [])) [step :end-a]))))

(defn b-step-fn [f step opts]
  (let [started (assoc opts k/steps-trace (conj (vec (get opts k/steps-trace [])) [step :start-b]))
        next (f step started)]
    (assoc next k/steps-trace (conj (vec (get next k/steps-trace [])) [step :end-b]))))

(deftest validates-step-fn-options
  (is (thrown-with-msg? IllegalArgumentException #"At least one" (core/create-step-fn {})))
  (is (thrown-with-msg? IllegalArgumentException #"before-f" (core/create-step-fn {:after-f :same}))))

(deftest threads-opts-and-wraps-step-fns-in-order
  (let [wf (core/create-workflow
            {:first-step :test/start
             :wire-fn (fn [step]
                        (case step
                          :test/start [(fn [opts] (assoc (core/ok opts) :test/bar "baz")) :test/end]
                          :test/end [identity nil]
                          [identity nil]))})
        res (wf [a-step-fn b-step-fn] {})]
    (is (= 0 (get res k/exit)))
    (is (nil? (get res k/err)))
    (is (= "baz" (:test/bar res)))
    (is (= [[:test/start :start-a]
            [:test/start :start-b]
            [:test/start :end-b]
            [:test/start :end-a]
            [:test/end :start-a]
            [:test/end :start-b]
            [:test/end :end-b]
            [:test/end :end-a]]
           (get res k/steps-trace)))))

(deftest captures-exceptions-into-opts
  (let [wf (core/create-workflow
            {:first-step :test/start
             :wire-fn (fn [step]
                        (if (= step :test/start)
                          [(fn [_opts] (throw (ex-info "boom" {:extra 1}))) :test/end]
                          [identity nil]))})
        res (wf [] {})]
    (is (= 1 (get res k/exit)))
    (is (= "boom" (get res k/err)))
    (is (= 1 (:extra res)))))
